### PR TITLE
feat: kafka trace consume

### PR DESCRIPTION
**Pull request by mabdinur**
Adds tracing and DSM support for https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#confluent_kafka.Consumer.consume

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Comment by wconti27** [comment]
Code looks good, my only concern is do we want to be creating a span per message? If we do, we'd probably want this configureable. If enabled, we create a span per message, else create one span for the whole batch. Looping in Teague on this one @tabgok

File: ddtrace/contrib/kafka/patch.py [hunk 1]
```diff
@@ -115,8 +115,11 @@ def patch():
     for producer in (TracedProducer, TracedSerializingProducer):
         trace_utils.wrap(producer, "produce", traced_produce)
     for consumer in (TracedConsumer, TracedDeserializingConsumer):
-        trace_utils.wrap(consumer, "poll", traced_poll)
-        trace_utils.wrap(consumer, "commit", traced_commit)
+        trace_utils.wrap(consumer, "poll", traced_poll_or_consume)
+        trace_utils.wrap(consumer, "consume", traced_poll_or_consume)
+
+    # Consume is not implemented in deserializing consumers
+    trace_utils.wrap(TracedConsumer, "commit", traced_commit)
```
Comment (mabdinur) [inline ddtrace/contrib/kafka/patch.py:122]:
```suggestion
        trace_utils.wrap(consumer, "commit", traced_poll_or_consume)

    # Consume is not implemented in deserializing consumers
    trace_utils.wrap(TracedConsumer, "consume", traced_commit)
```

File: ddtrace/contrib/kafka/patch.py [hunk 2]
```diff
@@ -115,8 +115,11 @@ def patch():
     for producer in (TracedProducer, TracedSerializingProducer):
         trace_utils.wrap(producer, "produce", traced_produce)
     for consumer in (TracedConsumer, TracedDeserializingConsumer):
-        trace_utils.wrap(consumer, "poll", traced_poll)
-        trace_utils.wrap(consumer, "commit", traced_commit)
+        trace_utils.wrap(consumer, "poll", traced_poll_or_consume)
+        trace_utils.wrap(consumer, "commit", traced_poll_or_consume)
+
+    # Consume is not implemented in deserializing consumers
+    trace_utils.wrap(TracedConsumer, "consume", traced_commit)
```
Comment (mabdinur) [inline ddtrace/contrib/kafka/patch.py:122]:
```suggestion
        trace_utils.wrap(consumer, "commit", traced_commit)

    # Consume is not implemented in deserializing consumers
    trace_utils.wrap(TracedConsumer, "consume", traced_poll_or_consume)
```

File: ddtrace/contrib/kafka/patch.py [hunk 3]
```diff
@@ -205,9 +211,23 @@ def traced_poll(func, instance, args, kwargs):
...
+    elif isinstance(result, list):
+        for message in result:
+            _instrument_message(message, pin, start_ns, err, instance)
+
+    if err is not None:
+        raise err
+    return result
+
+
+def _instrument_message(message, pin, start_ns, err, instance):
```
Comment (mabdinur) [inline ddtrace/contrib/kafka/patch.py:230]:
```suggestion
def _instrument_message(message, pin, start_ns, instance):
```

File: ddtrace/contrib/kafka/patch.py [hunk 4]
```diff
@@ -205,9 +211,23 @@ def traced_poll(func, instance, args, kwargs):
...
+        result = func(*args, **kwargs)
     except Exception as e:
         err = e
+
+    # consume returns a list of messages, poll returns a single message
+    if isinstance(result, confluent_kafka.Message):
+        _instrument_message(result, pin, start_ns, err, instance)
+    elif isinstance(result, list):
+        for message in result:
+            _instrument_message(message, pin, start_ns, err, instance)
```
Comment (mabdinur) [inline ddtrace/contrib/kafka/patch.py:223]:
```suggestion
            _instrument_message(message, pin, start_ns, instance)
```

File: ddtrace/contrib/kafka/patch.py [hunk 5]
```diff
@@ -136,6 +139,9 @@ def unpatch():
         if trace_utils.iswrapped(consumer.commit):
             trace_utils.unwrap(consumer, "commit")
 
+    # Consume is not implemented in deserializing consumers
+    trace_utils.unwrap(TracedConsumer, "consume")
```
Comment (mabdinur) [inline ddtrace/contrib/kafka/patch.py:143]:
```suggestion
    if trace_utils.iswrapped(TracedConsumer.consume)
        trace_utils.unwrap(TracedConsumer, "consume")
```

File: ddtrace/contrib/kafka/patch.py [hunk 6]
```diff
@@ -136,6 +139,10 @@ def unpatch():
         if trace_utils.iswrapped(consumer.commit):
             trace_utils.unwrap(consumer, "commit")
 
+    # Consume is not implemented in deserializing consumers
+    if trace_utils.iswrapped(TracedConsumer.consume)
+        trace_utils.unwrap(TracedConsumer, "consume")
```
Comment (mabdinur) [inline ddtrace/contrib/kafka/patch.py:144]:
```suggestion
    if trace_utils.iswrapped(TracedConsumer.consume)
        trace_utils.unwrap(TracedConsumer, "consume")
```

File: tests/contrib/kafka/test_kafka.py [hunk 7]
```diff
@@ -266,6 +266,48 @@ def test_commit(producer, consumer, kafka_topic):
...
+
+
+@pytest.mark.snapshot(ignores=["metrics.kafka.message_offset"])
+def test_commit_with_consume_with_error(producer, consumer, kafka_topic):
+    with override_config("kafka", dict(trace_empty_poll_enabled=True)):
+        producer.produce(kafka_topic, PAYLOAD, key=KEY)
+        producer.flush()
+        with pytest.raises(TypeError):
+            # tries to consume messages after the consumer has been closed with raises a RuntimeError
+            consumer.consume(num_messages=1, invalid_args="invalid_args")
```
Comment (mabdinur) [inline tests/contrib/kafka/test_kafka.py:308]: The kafka test suite filters out empty poll spans. So this consumer.consume(....) call generates a span but does not submit it to the agent. This is shown in the snapshot file. Not sure if we should fix this

File: releasenotes/notes/trace-kafka-consume-10a797a1305b1cd9.yaml [hunk 8]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    kafka: Adds tracing and DSM support for ``confluent_kafka.Consumer.consume()``. Previously only `confluent_kafka.Consumer.poll` was instrumented.
```
Comment (Yun-Kim) [inline releasenotes/notes/trace-kafka-consume-10a797a1305b1cd9.yaml:4]:
```suggestion
    kafka: Adds tracing and DSM support for ``confluent_kafka.Consumer.consume()``. Previously only `confluent_kafka.Consumer.poll` was instrumented.

```

**Review by Yun-Kim — APPROVED** [review]

File: ddtrace/contrib/kafka/patch.py [hunk 9]
```diff
@@ -204,10 +213,26 @@ def traced_poll(func, instance, args, kwargs):
...
+    finally:
+        # consume returns a list of messages, poll returns a single message
+        if result is None or isinstance(result, confluent_kafka.Message):
+            _instrument_message(result, pin, start_ns, instance, err)
+        elif isinstance(result, list):
+            messages = result
+            if len(result) > 1 and not config.kafka._batched_consumer_enabled:
+                messages = [result[0]]
+            for message in messages:
+                _instrument_message(message, pin, start_ns, instance, err)
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/contrib/kafka/patch.py:232]: Ahh I misunderstood our convo. So we want one span but dsm checkpoints for each message. I will make this change

[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/contrib/kafka/patch.py:232]: If we need to assign a DSM checkpoint for each message then we should also generate a span for each message. We should keep this functionality consistent with poll(...).

**Review by wconti27 — APPROVED** [review]
LGTM!

**Pull request merged**